### PR TITLE
[core] Resize string without no initialization

### PR DESCRIPTION
**Pull request by dentiny**
A super useful string util function.

Problem to solve: when `std::vector<char>` or `std::string` resizes, all appended memory will be filled in with initial value, which I frequently observe hotspotting when resize. 

How absl resize with no initialization works: depending on STL implementation (library type and version), some libraries have implemented an extra interface `resize_and_overwrite`, which doesn't memset for extra bytes.
The new interface is part of the standard starting from C++23: https://en.cppreference.com/w/cpp/string/basic_string/resize_and_overwrite

Example code place I think helpful:
https://github.com/ray-project/ray/blob/31b8b416f7c7f7345b1373bb1f45d14de60235f7/src/ray/common/client_connection.cc#L291-L292

### CODE REVIEW

File: src/ray/util/resize_uninitialized.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <string>
+
+#include "absl/strings/internal/resize_uninitialized.h"
```
Comment (jjyao) [inline src/ray/util/resize_uninitialized.h:19]: seems we are depending on absl private APIs?

**Comment by rynewang** [comment]
To save memset we can simply change `resize` to `reserve` and it can do the trick.

When I read into the code I realized the copy-to-vector should not be needed at all: `WriteMessageAsync` and `WriteMessage` only ever receives fbb so instead of passing a `int64_t length, const uint8_t *message,` we should really pass a `DetachedBuffer` from `fbb.Release()` and this whole copy is saved. Would you mind trying to do that?

**Comment by dentiny** [comment]
> To save memset we can simply change `resize` to `reserve` and it can do the trick.

That's complete different APIs!
`reserve` only reserves capacity while not changing size; while `resize` changes both.

File: src/ray/util/resize_uninitialized.h (hunk 1 above)
Comment (dentiny) [reply src/ray/util/resize_uninitialized.h:19]: This is the only interface I know which expose no memset functionality sadly

**Comment by dentiny** [comment]
> I know. but if the use case is `resize` then `assign`, we can change to `reserve` then `assign` with slightly better perf and no difference.

Actually `assign` is enough, no need to `reserve` or `resize`.

**Comment by dentiny** [comment]
I think there're some discussions on this PR, might take some time, let me make another PR for assignment only.

**Pull request closed without merging**